src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        stateCopy = { ...stateCopy, ...extraData };
+        break;
+
+      case 'removeProperties':
+        for (const item of keysToRemove) {
+          delete stateCopy[item];
+        }
+        break;
+
+      case 'clear':
+        for (const key in stateCopy) {
+          delete stateCopy[key];
+        }
```

Comment:
The clear case is currently deleting properties one by one, which is not necessary. You can simply assign an empty object to stateCopy to achieve the same result more efficiently.